### ISSUE TITLE
Adopt flat bundles?

### ISSUE THREAD

**Issue by robrichard** [facebook/relay#1590]
Are there plans to use rollup and generate flat bundles for Relay (similar to what react did in https://github.com/facebook/react/pull/9327)? It would be nice to get smaller bundles and it looks like this would also address the performance regression in server side rendering (https://github.com/facebook/relay/issues/1321).

**Comment by kassens** [comment]
We are a few days away from Making Travis Green Again and since the bundling/building is not used internally, this could be a great external contribution. I'll give @leebyron a chance to chime in who know the node landscape a bit better.

**Comment by gaearon** [comment]
FWIW we still `require()` fbjs modules in the React flat bundle. Is there any downside to this? I’m not sure I understand the discussion in https://github.com/facebook/relay/issues/1321.

**Comment by twhid** [comment]
Re: #1321 

Having the `require`s inline, particularly if you're using `require` hooks (eg babel/register), slows down a node server considerably. We saw a ~75% decrease in server rendering time when we switched to a custom relay build that hoists the `require` calls to the top of the scope.

**Comment by gaearon** [comment]
Oh I see. We do hoist them in our React flat bundles.
They look like:

```js
var invariant = require('fbjs/lib/invariant');
var warning = require('fbjs/lib/warning');

// all flattened modules below are using those
```

**Comment by leebyron** [comment]
This would be great! Right now we're using webpack to create these, but I agree that using Rollup would be better. I think this is a great task for someone looking to contribute to work on

**Comment by robrichard** [comment]
I opened a PR for this here: https://github.com/facebook/relay/pull/1610

**Comment by wincent** [comment]
> Right now we're using webpack to create these, but I agree that using Rollup would be better.

Haven't used it myself yet, but webpack (as of v2) also has [some support for tree-shaking features pioneered in Rollup](https://webpack.js.org/guides/tree-shaking/).

**Comment by YasserKaddour** [comment]
@wincent unfortunately tree shaking is broken in webpack, see https://github.com/webpack/webpack/issues/2867

**Comment by alloy** [comment]
Looks like webpack 4 is promising to fix this https://github.com/webpack/webpack/issues/2867#issuecomment-353714614. @robrichard is tree shaking the main motivation for this issue and linked PR?

**Comment by gaearon** [comment]
I don't think so. This is beneficial for many reasons. In our case the main benefits were caching `process.env.NODE_ENV` access in SSR, being able to inline more helpers using GCC simple mode, protecting internals from being imported. I wrote a post about this:

https://reactjs.org/blog/2017/12/15/improving-the-repository-infrastructure.html#compiling-flat-bundles

**Comment by alloy** [comment]
@gaearon That’s great info, thanks 👍 

@kassens Then I guess [the question remains](https://github.com/facebook/relay/pull/1610#issuecomment-361092126), is it possible to adopt ES6 modules in Relay from FB’s perspective?

**Comment by robrichard** [comment]
I'd like to start working on this again. I'm going to abandon the current (new very outdated) PR and start fresh. I'm very interested in getting this working as it would solve the server side rendering performance degradation described in https://github.com/facebook/relay/issues/1321#issuecomment-361308163 without degrading client side performance.

I think the most incremental approach would be separate PRs that do something like this:

1. Ensure all modules and types are imported using relative paths (complete what was started in https://github.com/facebook/relay/commit/97f24f8ebe80ebfd849db8e88f8449bfc9f9f5eb)
2. Publish the files in the same directory structure that they are in [/packages](https://github.com/facebook/relay/tree/master/packages) (instead of all files in `/lib` and remove the build transformation steps that rewrite the imports.
3. Update all `require`s to ESM imports/exports.
4. Update the build process to produce flat bundles using rollup.

@jstejada @alloy do you think this makes sense? Is it realistic for these changes to be merged? Would any of this interfere with how FB consumes this package?

**Comment by alloy** [comment]
I obviously can’t speak for FB, but the order of things makes sense to me 👍 

Changing to ESM would also eliminate [the one runtime addition](https://github.com/facebook/relay/pull/2293/files#diff-2c18fc91f11394e307550103afff9722R56) that my language plugin PR adds.

Might be a good idea to check if the jest-haste package is able to output all the paths of resolved haste modules?

**Comment by kassens** [comment]
@robrichard  the first 2 steps sound right. In fact, I think I already completed number 1.

2 makes sense to me and would allow us to clean that ugly transform rewriting requires. Would be an annoying, but not terrible breaking change for folks directly requireing the deeper modules, but seems fine.

3 we cannot do yet as we have an environment that only consumes requires and we use the Relay master verbatim there. What we could do is adopt Rollup first (there's some plugin to consume `require`s that I think React is using too?). After we have that, I can see if we can consume the bundled version internally (I hope it reduces overhead of the many modules!). If it should work internally and we just consume the flat bundle, we can do the last step and convert to `import`.

Hope that makes sense!

**Comment by josephsavona** [comment]
Testing

**Comment by josephsavona** [comment]
Test

**Comment by josephsavona** [comment]
Test

**Comment by josephsavona** [comment]
Test

**Comment by josephsavona** [comment]
Test!

**Comment by endiliey** [comment]
tes

**Comment by lolcoolkat** [comment]
LOL

**Comment by AugustoCalaca** [comment]
test comment

**Comment by ruslanvs** [comment]
Ha Ha

**Comment by ruslanvs** [comment]
asdfad

**Comment by ruslanvs** [comment]
asdfa

**Comment by ruslanvs** [comment]
asdfasdfasd

**Comment by ruslanvs** [comment]
asdfadf

**Comment by ruslanvs** [comment]
sadfa;djkh lkj

**Comment by ruslanvs** [comment]
32323

**Comment by ruslanvs** [comment]
asdf ad asdf

**Comment by ruslanvs** [comment]
wqefq r232

**Comment by ruslanvs** [comment]
asdfa asdf asdf

**Comment by ruslanvs** [comment]
asdfasdf

**Comment by ruslanvs** [comment]
asdfasd

**Comment by ruslanvs** [comment]
asdfasd

**Comment by ruslanvs** [comment]
111

**Comment by ruslanvs** [comment]
222

**Comment by ruslanvs** [comment]
asdfasd

**Comment by ruslanvs** [comment]
w3453sfgsd

**Comment by jstejada** [comment]
@ruslanvs please stop adding these comments

![image](https://user-images.githubusercontent.com/1271509/68913927-e8901900-072a-11ea-923f-54d833af66e4.png)

**Comment by ruslanvs** [comment]
@jstejada Sorry about that. I was working on a sample Relay app, and was actually wondering where it was persisting those comments...  Did not mean to spam you. Please feel free to delete this one and your message to me.

**Comment by liweinan0423** [comment]
s

**Comment by maxslayer77** [comment]
werf
